pkg/knuu/instance.go

```diff
@@ -626,49 +595,83 @@ func (i *Instance) SetStartupProbe(startupProbe *v1.Probe) error {
+func (i *Instance) AddSidecar(sidecar *Instance) error {
+
+	if !i.IsInState(Preparing, Committed) {
+		return fmt.Errorf("adding sidecar is only allowed in state 'Preparing' or 'Committed'. Current state is '%s'", i.state.String())
+	}
+	if sidecar == nil {
+		return fmt.Errorf("sidecar is nil")
+	}
+	if sidecar == i {
+		return fmt.Errorf("sidecar cannot be the same instance")
+	}
+	if sidecar.state != Committed {
+		return fmt.Errorf("sidecar '%s' is not in state 'Committed'", sidecar.name)
+	}
+	if i.isSidecar {
+		return fmt.Errorf("sidecar '%s' cannot have a sidecar", i.name)
+	}
+	if sidecar.isSidecar {
+		return fmt.Errorf("sidecar '%s' is already a sidecar", sidecar.name)
+	}
+
+	i.sidecars = append(i.sidecars, sidecar)
+	sidecar.isSidecar = true
+	sidecar.parentInstance = i
+	logrus.Debugf("Added sidecar '%s' to instance '%s'", sidecar.name, i.name)
+	return nil
+}
+
+// StartWithoutWait starts the instance without waiting for it to be ready
+// This function can only be called in the state 'Committed' or 'Stopped'
+func (i *Instance) StartWithoutWait() error {
 	if !i.IsInState(Committed, Stopped) {
-		return fmt.Errorf("starting is only allowed in state 'Committed'. Current state is '%s'", i.state.String())
+		return fmt.Errorf("starting is only allowed in state 'Committed' or 'Stopped'. Current state is '%s'", i.state.String())
 	}
-	if i.state == Committed {
-		if len(i.portsTCP) != 0 || len(i.portsUDP) != 0 {
-			logrus.Debugf("Ports not empty, deploying service for instance '%s'", i.k8sName)
-			svc, _ := k8s.GetService(k8s.Namespace(), i.k8sName)
-			if svc == nil {
-				err := i.deployService()
-				if err != nil {
-					return fmt.Errorf("error deploying service for instance '%s': %w", i.k8sName, err)
-				}
-			} else if svc != nil {
-				err := i.patchService()
-				if err != nil {
-					return fmt.Errorf("error patching service for instance '%s': %w", i.k8sName, err)
-				}
-			}
+	for _, sidecar := range i.sidecars {
+		if !sidecar.IsInState(Committed, Stopped) {
+			return fmt.Errorf("starting is only allowed in state 'Committed' or 'Stopped'. Current state of sidecar '%s' is '%s'", sidecar.name, sidecar.state.String())
 		}
-		if len(i.volumes) != 0 {
-			err := i.deployVolume()
-			if err != nil {
-				return fmt.Errorf("error deploying volume for instance '%s': %w", i.k8sName, err)
-			}
+	}
+	if i.isSidecar {
+		return fmt.Errorf("starting a sidecar is not allowed")
+	}
+	if i.state == Committed {
+
+		if err := i.deployResources(); err != nil {
+			return fmt.Errorf("error deploying resources for instance '%s': %w", i.k8sName, err)
 		}
-		if len(i.files) != 0 {
-			err := i.deployFiles()
-			if err != nil {
-				return fmt.Errorf("error deploying files for instance '%s': %w", i.k8sName, err)
+		for _, sidecar := range i.sidecars {
+			if err := sidecar.deployResources(); err != nil {
+				return fmt.Errorf("error deploying resources for sidecar '%s': %w", sidecar.k8sName, err)
 			}
```

Comment:
you can refactor the whole thing into a func that accepts a functions as a signature and ranges the given sidecar variable. This will eliminate most of the ranges and keep the code cleaner

Reply:
Yes! Fantastic. 
Still, I'd name it differently from performActions

Reply:
Renamed and refactored the function to be more generic: https://github.com/celestiaorg/knuu/pull/168/commits/9151c3e8f646e0fc661166cbcdbf030712de5209